textile/features.textile

```diff
@@ -739,7 +739,7 @@ h3(#realtime-presence). RealtimePresence
 ** @(RTP17a)@ All members belonging to the current connection are published as a @PresenceMessage@ on the @RealtimeChannel@ by the server irrespective of whether the client has permission to subscribe or the @RealtimeChannel@ is configured to publish presence events. A test should exist that attaches to a @RealtimeChannel@ with a @presence@ capability and without a @subscribe@ capability. It should then enter the @RealtimeChannel@ and ensure that the member entered from the current connection is present in the internal and public presence set available via "@RealtimePresence#get@":#RTP11
 ** @(RTP17b)@ The events that should be applied to the @RTP17@ presence map are: any @ENTER@, @PRESENT@ or @UPDATE@ event with a @connectionId@ that matches the current client's @connectionId@; any @LEAVE@ event with a @connectionId@ that matches the current client's @connectionId@ and is not a 'synthesized leave' (an event that has a connectionId which is not an initial substring of its id, per "@RTP2b1@":#RTP2b1 )
 ** @(RTP17h)@ Unlike the main @PresenceMap@, which is keyed by "memberKey":#TP3h , the @RTP17@ @PresenceMap@ must be keyed only by @clientId@. (Otherwise, entries associated with old @connectionId@s would never be removed, even if the user deliberately leaves presence).
-** @(RTP17f)@ The RealtimePresence object should perform automatic re-entry whenever a channel moves into the @ATTACHED@ state. (It does not need to do so for @RTL12@ @ATTACHED@ events received on already-@ATTACHED@ channels).
+** @(RTP17f)@ The RealtimePresence object should perform automatic re-entry whenever a channel moves into the @ATTACHED@ state. (It does not need to do so for @RTL12@ @ATTACHED@ events received on already-@ATTACHED@ channels which were resumed).
```

Comment:
>It still says "should perform automatic re-entry whenever a channel moves into the ATTACHED state"

But I've updated the part that says "already-@ATTACHED@".

>Please refer to "Modification"

Sure, sorry, it's not just a patch.

Reply:
I think that the phrase "moves into" is too vague here; seems like different people are interpreting it differently. To me, "moves into the `ATTACHED` state" means that the channel transitions _from a non-`ATTACHED` state_ to `ATTACHED`. If that's the case, then a direct implication of "should perform automatic re-entry whenever a channel moves into the `ATTACHED` state" is that _no_ RTL12 `ATTACHED` event should trigger an automatic re-entry (since the channel is already attached in the case of one of these events). And my reading of the original RTP17f text was that the parenthesised sentence was simply pointing out this implication. If the intention of the parenthesised sentence is to actually refine the behaviour that the spec point is defining, then I think the wording of the whole spec point needs to be clearer.

Reply:
2d4c50b916ca775a2bda2223e1685faaef81c456

Reply:
https://github.com/ably/specification/pull/182/commits/f9cbbed9bac0be6fecf1c16e58324b75c05e41f8

Reply:
> Are we now entering presence for every type of attach?

Yes, "whenever a channel receives an @ATTACHED@ protocolmessage".

Reply:
https://github.com/ably/ably-go/blob/dd8236fc301eb078803e83efee7615f3ba133341/ably/realtime_channel.go#L803-L811

Reply:
> We have implemented code in ably-go code as per ...

yes? you wanted a simpler and more easily understandable/implementable spec item, this was my suggestion

Reply:
> Lawrence is correct. This is (a) a behaviour change and (b) the wording makes it ambiguous when it wasn't before.
> 
> If we're going to go to the bother of making this change let's do it in a way that simplifies things. Eg remove all references to the state and just say `The RealtimePresence object should perform automatic re-entry whenever a channel receives an @ATTACHED@ protocolmessage` or something.

>this was my suggestion

Is there a PR with this suggestion?

Reply:
>you wanted a simpler and more easily understandable/implementable spec item

Do you mind if it'll remain as @sacOO7 is proposing though? "(For ATTACHED events received on already-ATTACHED channels, it should only do it for resumed flag set to false)"

Reply:
>you wanted a simpler and more easily understandable/implementable spec item

@SimonWoolf Do you mind if it'll remain as @sacOO7 is proposing though? "(For ATTACHED events received on already-ATTACHED channels, it should only do it for resumed flag set to false)"

Reply:
@sacOO7 The situation when library sends re-enter with resumed=true is rare and not worth complications both in the spec and the code. WDYT? Maybe we will go with it? It's not very hard to change this in ably-go as I can imagine.

Reply:
Hi @maratal . I know it's a rare event but we should keep it explicit and simple. I don't agree with both cases being right as such because second case can happen ( when there is a message loss due to account limits or other reasons ).
I feel if second case feels simple we should keep it. If it still feels complicated, we can go with first one. I feel, we have pretty much simplified second case WDYT

Reply:
Ok, lets go with the second one. My wording will be this:
`The RealtimePresence object should perform automatic re-entry whenever a channel receives an @ATTACHED@ protocol message, except those which were received on an already ATTACHED channel with a resumed flag set to true.` WDYT @sacOO7 @SimonWoolf @lawrence-forooghian

Reply:
I think we can keep conditions assertive. 
`The RealtimePresence object should perform automatic re-entry whenever ATTACHED event received. For ATTACHED event received on already-@ATTACHED@ channel, it should only do it for resumed flag set to false`

Reply:
LGTM

Reply:
Sachin's wording doesn't make sense — you can't say "do x _whenever_ y happens" and then in the next sentence say "sometimes, when y happens, don't do x". I suggest `The @RealtimePresence@ object should perform automatic re-entry whenever a channel receives an @ATTACHED@ @ProtocolMessage@, except in the case where the channel is already attached and the @ProtocolMessage@ has the @RESUMED@ bit flag set.`.

Reply:
Slightly shorter version: `The @RealtimePresence@ object should perform automatic re-entry whenever the channel receives an @ATTACHED@ @ProtocolMessage@, except for an already attached channels where received @ProtocolMessage@ has the @RESUMED@ bit flag set.` wdyt @lawrence-forooghian @sacOO7

Reply:
I think what I wrote is more readable 🤷

Reply:
I believe bc I'm not native speaker 🙂

Reply:
I think we are trying to write the spec in away we have implemented it. I think in case of cocoa, we are doing early return where as in other SDK's we take assertive condition and execute presence enter.

Reply:
@maratal I am okay to go with last version as long as it feels understandable to you.

Reply:
`The @RealtimePresence@ object should perform automatic re-entry whenever the channel receives an @ATTACHED@ @ProtocolMessage@, except for an already attached channels where received @ATTACHED@@ProtocolMessage@ has @RESUMED@ bit flag set to true`

Reply:
@lawrence-forooghian Looks like @sacOO7 is more comfortable with my variant, I'm good with both of them, would you mind if we go further with mine?

Reply:
I mean, I worry that starting to go back and forth on wording is not a particularly productive use of time, but as I said I think what I wrote reads better.

Reply:
Do you believe that there is a difference in meaning between the two versions?

Reply:
I'm good to go with either way : )

Reply:
https://github.com/ably/specification/pull/182/commits/be4bbb6baaadd0958fb8c9d818bbd6acb01f3224
@lawrence-forooghian @sacOO7 @SimonWoolf please approve if you are fine with what we all ended up with. Thanks.

Reply:
Resolving it myself due to some availability bottleneck.